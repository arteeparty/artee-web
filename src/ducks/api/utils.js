import cache from "memory-cache";

const API_BASE = process.env.REACT_APP_API_BASE || "https://api.artee.party";

export function resolveURL(path, scheme = "http") {
  if (scheme !== "http") {
    return API_BASE.replace("http", scheme === "wss" ? "ws" : scheme) + path;
  }
  return API_BASE + path;
}

const handleResponse = async (url, response, time = 0, method = "GET") => {
  let output;
  if (
    response.headers.get("Content-Type") === "application/json; charset=utf-8"
  ) {
    output = await response.json();
  } else {
    output = await response.text();
  }
  if (!response.ok) {
    const error = {
      status: response.status,
      message: output.message || response.statusText
    };
    throw error;
  }
  if (time !== 0) {
    cache.put(`${method} | ${url}`, output, time);
  }
  return output;
};

export const fetchAPI = async (route, config) => {
  const url = resolveURL(route);

  let cachedResponse = cache.get(`${config.method || "GET"} | ${url}`);
  if (cachedResponse) {
    return cachedResponse;
  }
  const options = {
    ...config,
    headers: {
      ...config.headers
    }
  };
  if (options.body) {
    options.body = JSON.stringify(options.body);
    options.headers["Content-Type"] = "application/json";
  }
  const response = await fetch(url, options);
  return handleResponse(url, response, config.cache || 0, config.method);
};
