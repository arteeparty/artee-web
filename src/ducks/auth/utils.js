import cookies from "js-cookie";

export function loadAuthState() {
  const { access_token, refresh_token } = cookies.getJSON("sid") || {
    access_token: null,
    refresh_token: null
  };
  return {
    profile: cookies.getJSON("profile") || null,
    access_token,
    refresh_token
  };
}

export function setAuthState(access_token, refresh_token, profile) {
  cookies.set(
    "sid",
    {
      access_token,
      refresh_token
    },
    { secure: false }
  );
  if (profile) cookies.set("profile", profile, { secure: false });
}
