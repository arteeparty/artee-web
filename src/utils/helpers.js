export const createActionTypes = (prefix, actionList) => {
  return actionList.reduce((actions, action) => {
    actions[action] = `${prefix}/${action}`;
    return actions;
  }, {});
};

export function generateUUID(a, b) {
  for (
    b = a = "";
    a++ < 36;
    b +=
      (a * 51) & 52
        ? (a ^ 15 ? 8 ^ (Math.random() * (a ^ 20 ? 16 : 4)) : 4).toString(16)
        : "-"
  );
  return b;
}

export function join(/* path segments */) {
  // Split the inputs into a list of path commands.
  var parts = [];
  for (var i = 0, l = arguments.length; i < l; i++) {
    parts = parts.concat(arguments[i].toString().split("/"));
  }
  // Interpret the path commands to get the new resolved path.
  var newParts = [];
  for (i = 0, l = parts.length; i < l; i++) {
    var part = parts[i];
    // Remove leading and trailing slashes
    // Also remove "." segments
    if (!part || part === ".") continue;
    // Interpret ".." to pop the last segment
    if (part === "..") newParts.pop();
    else
      // Push new path segments.
      newParts.push(part);
  }
  // Preserve the initial slash if there was one.
  if (parts[0] === "") newParts.unshift("");
  // Turn back into a single string path.
  return newParts.join("/") || (newParts.length ? "/" : ".");
}

export const getInWithPath = (state, first, ...rest) => {
  if (!state) {
    return undefined;
  }
  const next = state[first];
  return rest.length ? getInWithPath(next, ...rest) : next;
};

export const getIn = (state, field) =>
  Array.isArray(field)
    ? getInWithPath(state, ...field)
    : getInWithPath(state, field);

// Adapted from https://github.com/erikras/redux-form/blob/v6/src/structure/plain/setIn.js
const setInWithPath = (state, value, first, ...rest) => {
  if (first === undefined) {
    return value;
  }
  const next = setInWithPath(state && state[first], value, ...rest);
  if (!state) {
    const initialized = isNaN(first) ? {} : [];
    initialized[first] = next;
    return initialized;
  }
  if (Array.isArray(state)) {
    const copy = [...state];
    copy[first] = next;
    return copy;
  }
  return {
    ...state,
    [first]: next
  };
};

export const setIn = (state, field, value) =>
  Array.isArray(field)
    ? setInWithPath(state, value, ...field)
    : setInWithPath(state, value, [field]);
