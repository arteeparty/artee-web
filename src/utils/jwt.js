import moment from "moment";
import jwtDecode from "jwt-decode";

export const getExpiration = token => {
  const { exp } = jwtDecode(token);
  const then = moment(exp * 1000);
  const now = moment();
  return then.diff(now, "milliseconds");
};
