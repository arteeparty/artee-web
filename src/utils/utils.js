const API_BASE = process.env.REACT_APP_API_BASE || "http://localhost:8001";

export function resolveURL(path, scheme = "http") {
  if (scheme !== "http") {
    return API_BASE.replace("http", scheme) + path;
  }
  return API_BASE + path;
}

const handleResponse = async (url, response, method = "GET") => {
  let output;
  if (response.headers.get("Content-Type") === "application/json") {
    output = await response.json();
  } else {
    output = await response.text();
  }
  if (!response.ok) {
    const error = {
      status: response.status,
      message: output.message || response.statusText
    };
    throw error;
  }
  return output;
};

export const fetchAPI = async (route, config) => {
  const url = resolveURL(route);

  const options = {
    ...config,
    headers: {
      ...config.headers
    }
  };
  if (options.body) {
    options.body = JSON.stringify(options.body);
    options.headers["Content-Type"] = "application/json";
  }
  const response = await fetch(url, options);
  return handleResponse(url, response, config.method);
};
