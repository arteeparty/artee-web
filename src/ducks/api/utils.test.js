import { fetchAPI } from "./utils";

const mockResponse = (
  status,
  statusText,
  response,
  headers = { "Content-Type": "application/json" }
) => {
  const h = Object.keys(headers).reduce((n, h) => {
    n.append(h, headers[h]);
    return n;
  }, new Headers());
  return new window.Response(response, {
    status: status,
    statusText: statusText,
    headers: h
  });
};

describe("fetchAPI", () => {
  it("should handle no method", async () => {
    window.fetch = jest.fn().mockImplementation((route, p) => {
      expect(route).toEqual("http://localhost:8000/route/cached");
      return Promise.resolve(mockResponse(200, null, '{"key": true}'));
    });
    let response = await fetchAPI("/route/cached", { cache: 2500 });
    expect(response).toEqual({ key: true });
    response = await fetchAPI("/route/cached", { cache: 2500 });
    expect(response).toEqual({ key: true });
  });

  it("should handle plain text body", async () => {
    window.fetch = jest.fn().mockImplementation((route, p) => {
      expect(route).toEqual("http://localhost:8000/route");
      return Promise.resolve(
        mockResponse(200, null, "fish", { "Content-Type": "text/plain" })
      );
    });
    const response = await fetchAPI("/route", { body: { test: true } });
    expect(response).toEqual("fish");
  });

  it("should handle failure", () => {
    window.fetch = jest.fn().mockImplementation((route, p) => {
      expect(route).toEqual("http://localhost:8000/route");
      return Promise.resolve(
        mockResponse(404, "not found", "fish", { "Content-Type": "text/plain" })
      );
    });
    const response = fetchAPI("/route", { body: { test: true } });
    expect(response).rejects.toMatchSnapshot();
    expect(response).rejects.toEqual({ message: "not found", status: 404 });
  });
});
