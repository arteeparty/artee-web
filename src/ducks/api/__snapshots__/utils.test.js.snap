// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAPI should handle failure 1`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).toEqual([22m[32mexpected[39m[2m)[22m

Expected value to equal:
  [32m"http://localhost:8000/route"[39m
Received:
  [31m"http://localhost:8001/route"[39m]
`;
